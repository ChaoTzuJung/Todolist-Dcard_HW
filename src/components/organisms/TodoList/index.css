.todolist {
	margin-top: 24px;
	& > .input-panel {
		margin-bottom: 24px;
	}

	@media screen and (--phone) {
		width: 100%;
	}
}
