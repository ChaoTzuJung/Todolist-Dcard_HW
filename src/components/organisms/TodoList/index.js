/* eslint-disable no-unused-vars */
import React, { Component } from 'react';
import classnames from 'classnames';
import Input from 'components/atoms/Input';
import TodoItem from 'components/molecules/TodoItem';
import { DragDropContext, Droppable, Draggable } from 'react-beautiful-dnd';

import { firebaseTodos, firebaseLooper, firebaseSort } from '../../../../firebase';
import styles from './index.css';

// a little function to help us with reordering the result
const reorder = (list, startIndex, endIndex) => {
	const result = Array.from(list);
	const [removed] = result.splice(startIndex, 1);
	result.splice(endIndex, 0, removed);
	return result;
};

class TodoList extends Component {
	constructor(props) {
		super(props);
		this.input = React.createRef();

		this.state = {
			todos: [],
			isNewTodo: false,
		};

		this.onFocus = this.onFocus.bind(this);
		this.handleNewTodo = this.handleNewTodo.bind(this);
		this.onDragEnd = this.onDragEnd.bind(this);
	}

	componentDidMount() {
		console.log('**TodoList** componentDidMount 嘍 !!');
		firebaseTodos.on('value', todoSnapshot => {
			// 確認Todo資料庫有資料 true or null
			if (todoSnapshot.val()) {
				// 有資料就抓出排序 讓 firebaseTodos 可以排序一下
				// eslint-disable-next-line no-shadow
				firebaseSort.once('value', sortSnapshot => {
					// 再確認Sort資料庫有資料 true or null
					if (sortSnapshot.val()) {
						// eslint-disable-next-line no-shadow
						firebaseSort.on('value', snapshot => {
							const sortIndex = [];
							const sortedArray = [];
							snapshot.forEach(childSnapshot => {
								// 抓出排序給 firebaseTodos 做參考
								sortIndex.push(childSnapshot.val());
							});
							// eslint-disable-next-line no-shadow
							firebaseTodos.once('value').then(snapshot => {
								const todoObject = snapshot.val();
								// 把 firebaseSort 排序塞入 firebaseTodo
								sortIndex.map(key => sortedArray.push(todoObject[key]));
								this.setState(prevState => ({
									...prevState.todos,
									todos: sortedArray,
								}));
							});
						});
					} else {
						// Sort 若無資料，直接把顯有 firebaseTodos 加入 state 並 渲染資料
						const todos = firebaseLooper(todoSnapshot);
						this.setState(prevState => ({
							...prevState.todos,
							todos,
						}));
					}
				});
			}
		});
	}

	onDragEnd(result) {
		// dropped outside the list
		if (!result.destination) {
			return;
		}
		const { todos } = this.state;
		const todo = reorder(todos, result.source.index, result.destination.index);
		// this.setState({
		// 	todos: todo,
		// });
		// 產生新順序
		const sortId = [];
		todo.forEach(obj => {
			sortId.push(obj.id);
		});

		// Todo 依照順序並渲染data跟畫面
		firebaseTodos.on('value', snapshot => {
			const TodoObj = snapshot.val();
			const answer = [];

			sortId.forEach(key => {
				let found = false;
				Object.keys(TodoObj).filter(item => {
					if (!found && item === key) {
						const newTodo = { ...TodoObj[item], id: item };
						answer.push(newTodo);
						found = true;
						return false;
					}
					return true;
				});
			});
			this.setState(prevState => ({
				...prevState.todos,
				todos: answer,
			}));
			// 建立順序
			firebaseSort.set(sortId);
		});
	}

	onFocus() {
		this.setState({
			isNewTodo: true,
		});
	}

	handleNewTodo() {
		this.setState({
			isNewTodo: false,
		});
	}

	render() {
		const { className, value, tab, ...props } = this.props;
		const { todos, isNewTodo } = this.state;
		return (
			<DragDropContext onDragEnd={this.onDragEnd}>
				<div className={classnames(styles.todolist, className)}>
					<div className={styles.inputPanel}>
						{isNewTodo ? (
							<TodoItem isNewTodo={isNewTodo} setNewTodo={this.handleNewTodo} />
						) : (
							<Input
								className={styles.inputButton}
								ref={this.input}
								onFocus={this.onFocus}
								placeholder="Add Task"
								{...props}
							/>
						)}
					</div>
					<Droppable droppableId="droppable">
						{/* eslint-disable-next-line no-unused-vars */}
						{(provided, snapshot) => (
							<div {...provided.droppableProps} ref={provided.innerRef}>
								{tab === 'myTasks' &&
									todos.map((todo, index) => (
										<Draggable key={todo.id} draggableId={todo.id} index={index}>
											{/* eslint-disable-next-line no-shadow */}
											{(provided, snapshot) => (
												<div
													ref={provided.innerRef}
													{...provided.draggableProps}
													{...provided.dragHandleProps}
												>
													<TodoItem
														key={todo.id}
														id={todo.id}
														message={todo.message}
														star={todo.star}
														date={todo.date}
														file={todo.file}
														name={todo.name}
														type={todo.type}
														comment={todo.comment}
														completed={todo.completed}
														isNewTodo={isNewTodo}
														setNewTodo={this.handleNewTodo}
													/>
												</div>
											)}
										</Draggable>
									))}
								{tab === 'inProgress' &&
									todos
										.reduce((filtered, todo) => {
											if (!todo.completed) {
												const filteredTodo = { ...todo };
												filtered.push(filteredTodo);
											}
											return filtered;
										}, [])
										.map((todo, index) => (
											<Draggable key={todo.id} draggableId={todo.id} index={index}>
												{/* eslint-disable-next-line no-shadow */}
												{(provided, snapshot) => (
													<div
														ref={provided.innerRef}
														{...provided.draggableProps}
														{...provided.dragHandleProps}
													>
														<TodoItem
															key={todo.id}
															id={todo.id}
															message={todo.message}
															star={todo.star}
															date={todo.date}
															file={todo.file}
															name={todo.name}
															type={todo.type}
															comment={todo.comment}
															completed={todo.completed}
															isNewTodo={isNewTodo}
															setNewTodo={this.handleNewTodo}
														/>
													</div>
												)}
											</Draggable>
										))}
								{tab === 'completed' &&
									todos
										.reduce((filtered, todo) => {
											if (todo.completed) {
												const filteredTodo = { ...todo };
												filtered.push(filteredTodo);
											}
											return filtered;
										}, [])
										.map((todo, index) => (
											<Draggable key={todo.id} draggableId={todo.id} index={index}>
												{/* eslint-disable-next-line no-shadow */}
												{(provided, snapshot) => (
													<div
														ref={provided.innerRef}
														{...provided.draggableProps}
														{...provided.dragHandleProps}
													>
														<TodoItem
															key={todo.id}
															id={todo.id}
															message={todo.message}
															star={todo.star}
															date={todo.date}
															file={todo.file}
															name={todo.name}
															type={todo.type}
															comment={todo.comment}
															completed={todo.completed}
															isNewTodo={isNewTodo}
															setNewTodo={this.handleNewTodo}
														/>
													</div>
												)}
											</Draggable>
										))}
								{provided.placeholder}
							</div>
						)}
					</Droppable>
				</div>
			</DragDropContext>
		);
	}
}

export default TodoList;
