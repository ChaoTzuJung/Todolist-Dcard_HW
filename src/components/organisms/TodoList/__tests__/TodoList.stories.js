import React from 'react';
import { withKnobs } from '@storybook/addon-knobs/react';

import TodoList from 'components/organisms/TodoList';

import { storiesOf } from '@storybook/react';

const stories = storiesOf('Commons|organisms/TodoList', module);

stories.addDecorator(withKnobs);

stories.add('__interactive', () => (
	<div style={{ margin: '100px' }}>
		<TodoList />
	</div>
));
