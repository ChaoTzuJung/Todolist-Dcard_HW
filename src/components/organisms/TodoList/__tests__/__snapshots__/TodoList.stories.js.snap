// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`__interactive 1`] = `
<div
  style={
    Object {
      "margin": "100px",
    }
  }
>
  <TodoList>
    <div
      className="todolist"
    >
      <Contain(Input)
        onFocus={[Function]}
        onSubmit={[Function]}
        placeholder="Add Task"
      >
        <Input
          content={Array []}
          getInputElement={[Function]}
          handleChange={[Function]}
          onFocus={[Function]}
          placeholder="Add Task"
          submit={[Function]}
          value=""
        >
          <form
            action=""
            autoComplete="off"
            className="input"
            onSubmit={[Function]}
          >
            <div
              className="icon"
            >
              <FontAwesomeIcon
                border={false}
                className=""
                color="#CBCBCB"
                fixedWidth={false}
                flip={null}
                icon={
                  Object {
                    "icon": Array [
                      448,
                      512,
                      Array [],
                      "f067",
                      "M416 208H272V64c0-17.67-14.33-32-32-32h-32c-17.67 0-32 14.33-32 32v144H32c-17.67 0-32 14.33-32 32v32c0 17.67 14.33 32 32 32h144v144c0 17.67 14.33 32 32 32h32c17.67 0 32-14.33 32-32V304h144c17.67 0 32-14.33 32-32v-32c0-17.67-14.33-32-32-32z",
                    ],
                    "iconName": "plus",
                    "prefix": "fas",
                  }
                }
                inverse={false}
                listItem={false}
                mask={null}
                pull={null}
                pulse={false}
                rotation={null}
                size="1x"
                spin={false}
                symbol={false}
                title=""
                transform={null}
              >
                <svg
                  aria-hidden="true"
                  className="svg-inline--fa fa-plus fa-w-14 fa-1x "
                  color="#CBCBCB"
                  data-icon="plus"
                  data-prefix="fas"
                  focusable="false"
                  role="img"
                  style={Object {}}
                  viewBox="0 0 448 512"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M416 208H272V64c0-17.67-14.33-32-32-32h-32c-17.67 0-32 14.33-32 32v144H32c-17.67 0-32 14.33-32 32v32c0 17.67 14.33 32 32 32h144v144c0 17.67 14.33 32 32 32h32c17.67 0 32-14.33 32-32V304h144c17.67 0 32-14.33 32-32v-32c0-17.67-14.33-32-32-32z"
                    fill="currentColor"
                    style={Object {}}
                  />
                </svg>
              </FontAwesomeIcon>
            </div>
            <input
              autoComplete="off"
              onChange={[Function]}
              onFocus={[Function]}
              placeholder="Add Task"
              role="searchbox"
              tabIndex="0"
              type="text"
              value=""
            />
          </form>
        </Input>
      </Contain(Input)>
    </div>
  </TodoList>
</div>
`;
